src/Disks/S3/DiskS3.cpp

```diff
@@ -63,332 +63,339 @@ void DiskS3::AwsS3KeyKeeper::addKey(const String & key)
             assertChar('\t', buf);
-            readIntText(total_size, buf);
-            assertChar('\n', buf);
-            s3_objects.resize(s3_objects_count);
-            for (UInt32 i = 0; i < s3_objects_count; ++i)
+            readEscapedString(s3_object_path, buf);
+            if (version == VERSION_ABSOLUTE_PATHS)
             {
-                String s3_object_path;
-                size_t s3_object_size;
-                readIntText(s3_object_size, buf);
-                assertChar('\t', buf);
-                readEscapedString(s3_object_path, buf);
-                if (version == VERSION_ABSOLUTE_PATHS)
-                {
-                    if (!boost::algorithm::starts_with(s3_object_path, s3_root_path))
-                        throw Exception(
-                            "Path in metadata does not correspond S3 root path. Path: " + s3_object_path
-                            + ", root path: " + s3_root_path + ", disk path: " + disk_path_,
-                            ErrorCodes::UNKNOWN_FORMAT);
-                    s3_object_path = s3_object_path.substr(s3_root_path.size());
-                }
-                assertChar('\n', buf);
-                s3_objects[i] = {s3_object_path, s3_object_size};
+                if (!boost::algorithm::starts_with(s3_object_path, s3_root_path))
+                    throw Exception(
+                        "Path in metadata does not correspond S3 root path. Path: " + s3_object_path
+                        + ", root path: " + s3_root_path + ", disk path: " + disk_path_,
+                        ErrorCodes::UNKNOWN_FORMAT);
+                s3_object_path = s3_object_path.substr(s3_root_path.size());
             }
-
-            readIntText(ref_count, buf);
             assertChar('\n', buf);
-
-            if (version >= VERSION_READ_ONLY_FLAG)
-            {
-                readBoolText(read_only, buf);
-                assertChar('\n', buf);
-            }
+            s3_objects[i] = {s3_object_path, s3_object_size};
         }
 
-        void addObject(const String & path, size_t size)
+        readIntText(ref_count, buf);
+        assertChar('\n', buf);
+
+        if (version >= VERSION_READ_ONLY_FLAG)
         {
-            total_size += size;
-            s3_objects.emplace_back(path, size);
+            readBoolText(read_only, buf);
+            assertChar('\n', buf);
         }
+    }
 
-        /// Fsync metadata file if 'sync' flag is set.
-        void save(bool sync = false)
-        {
-            WriteBufferFromFile buf(disk_path + metadata_file_path, 1024);
+    void addObject(const String & path, size_t size)
+    {
+        total_size += size;
+        s3_objects.emplace_back(path, size);
+    }
 
-            writeIntText(VERSION_RELATIVE_PATHS, buf);
-            writeChar('\n', buf);
+    /// Fsync metadata file if 'sync' flag is set.
+    void save(bool sync = false)
+    {
+        WriteBufferFromFile buf(disk_path + metadata_file_path, 1024);
```

Comment:
We should do:
- write to tmp file;
- fsync if need;
- rename;

Reply:
Metadata files are already written through tmp files or directories as DiskS3 completely repeats FS layout as in the case of DiskLocal.
The main issue is to correctly remove them from FS if they can't be parsed.